keps/sig-node/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1048 @@
+may fail to schedule. In practice, a workload that needs a GPU can be written
+such that it can discover the GPUs available to it, and work with what it is
+given. A user may have a preference for the latest model, but would like to run
+the workload even if only an older model is available.
+
+Similarly, packaged workload authors may wish to configure a workload such that
+it will work well in the widest selection of available clusters. That is, a
+distributor of shared workload definitions would like to be able to specify
+alternative types of devices with which their workload will function, without
+requiring the user to modify the manifests.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Allow workload authors, when specifying a `ResourceClaim`, to provide a list
+  of ways to satisfy the claim, with a preference ranking.
+* Enable schedulers to evaluate those preferences and allocate devices for the
+  claim based on them.
+* Enable cluster autoscalers to evaluate those preferences and make scaling
+  choices based on them.
+* Provide some measure of ResourceQuota controls when users utilize claims with
+  these types of requests.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Enable cross-claim consistency of request choices. For example, guaranteeing
+  that all `ResourceClaim`s associated with a given `Deployment` are satisfied
+  using the same choice from the list of possible alternatives.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal adds a new field to the `DeviceRequest`, called `FirstOf` which
+will contain an ordered list of `DeviceRequest` objects. In order to satisfy the
+main (containing) request, exactly one of the requests listed in `FirstOf` must
+be satisfied. The order listed is considered a priority order, such that the
+scheduler will only try to use the second item in the list if it is unable to
+satsify the first item, and so on.
+
+A `DeviceRequest` that populates the `FirstOf` field must *not* populate the
+`DeviceClassName` field. The `required` validation on this field will be
+relaxed. This allows existing clients to differentiate between claims they
+understand (with `DeviceClassName`) and those they do not (without
+`DeviceClassName` but with the new field). Clients written for 1.31, when
+`DeviceClassName` was required, were requested to include this logic, and the
+in-tree components have been built in this way.
+
+```go
```

Comment:
Here's the diff, if it's useful during review:

```diff
diff --git a/pkg/apis/resource/types.go b/pkg/apis/resource/types.go
index 2c03a304fdf..93b96c3e4c8 100644
--- a/pkg/apis/resource/types.go
+++ b/pkg/apis/resource/types.go
@@ -397,7 +397,11 @@ type DeviceRequest struct {
        // additional configuration and selectors to be inherited by this
        // request.
        //
-       // A class is required. Which classes are available depends on the cluster.
+       // Either a class or FirstOf requests are required in DeviceClaim.Requests. When
+       // this request is part of the FirstOf list, a class is required. Nested FirstOf
+       // requests are not allowed
+       //
+       // Which classes are available depends on the cluster.
        //
        // Administrators may use this to restrict which devices may get
        // requested by only installing classes with selectors for permitted
@@ -405,9 +409,19 @@ type DeviceRequest struct {
        // then administrators can create an empty DeviceClass for users
        // to reference.
        //
-       // +required
+       // +optional
+       // +oneOf=deviceRequestType
        DeviceClassName string
 
+       // FirstOf contains subrequests, exactly one of which must be satisfied
+       // in order to satisfy this request. This field may only be set in the
+       // entries of DeviceClaim.Requests. It must not be set in DeviceRequest
+       // instances that themselves are part of a FirstOf.
+       //
+       // +optional
+       // +oneOf=deviceRequestType
+       FirstOf []DeviceRequest
+
        // Selectors define criteria which must be satisfied by a specific
        // device in order for that device to be considered for this
        // request. All selectors must be satisfied for a device to be
@@ -457,7 +471,8 @@ type DeviceRequest struct {
 }
 
 const (
-       DeviceSelectorsMaxSize = 32
+       DeviceSelectorsMaxSize      = 32
+       FirstOfDeviceRequestMaxSize = 8
 )

```

Reply:
Some alternatives: `FirstAvailableFrom`, `FirstMatchFrom` (not feeling strongly for any particular one but leaving them here to see if folks have thoughts..)

Reply:
Latest version if `FirstAvailableOf` but happy to change...but I think it can also be done at implementation time, we don't need the exact API now.